### ISSUE TITLE
`flutter clean` doesn't clean everything

### ISSUE THREAD

**Issue by tvolkert** [flutter/flutter#12631]
A more general case of https://github.com/flutter/flutter/issues/8656

After running an app on iOS and then cleaning, there are still the following artifacts:

```
	ios/Flutter/App.framework/
	ios/Flutter/Flutter.framework/
	ios/Flutter/Generated.xcconfig
	ios/Flutter/app.flx
	ios/ServiceDefinitions.json
```

Ideally, we'd produce those artifacts in the `build/` directory, but in absence of that, we should at least clean them out.

**Comment by eseidelGoogle** [comment]
We presumably could write an automated test for this relatively easily, no?

**Comment by tvolkert** [comment]
Yeah, we should be able `flutter create`, record the files that exist, then `flutter build <options>` and diff the list of files against what we recorded.

**Comment by jonahwilliams** [comment]
I think @jmagman addressed most of the ios/xcode issues recently

**Comment by jmagman** [comment]
> I think @jmagman addressed most of the ios/xcode issues recently

As of https://github.com/flutter/flutter/pull/38992 the Xcode workspace is cleaned during a `flutter clean` but the files @tvolkert mentioned are not deleted.  

I'm not sure if `ios/Flutter/Flutter.framework/` can be deleted without requiring running `pod install` to re-copy the right build version, and I don't think the logic for whether that command is run checks for the presence of that bundle.  That's fixable, but I want to leave a bread-crumb trail for any engineers who pick up this issue.

TL;DR make sure `flutter build ios && flutter clean && flutter build ios` works both with and without iOS platform plugins.

**Comment by jmagman** [comment]
Removed 
```
ios/Flutter/App.framework/
ios/Flutter/Generated.xcconfig
ios/Flutter/flutter_export_environment.sh
```
with https://github.com/flutter/flutter/pull/51532.

Keeping this open to track what's left.

**Comment by xster** [comment]
I suspect we can clean out Flutter.framework too. Otherwise, if you have 20 Flutter projects and you flutter ran them once each, you'd have 20 copies of various Flutter engines lying around on your disk.

**Comment by jmagman** [comment]
@xster That would definitely be more work.  Flutter.framework needs to linkable in a known directory (wherever the Podfile says it is) for plugins to build, which happens before the Runner target builds and `xcode_backend` can run to copy the right version of the engine around.  See https://github.com/flutter/flutter/pull/42029.

**Comment by xster** [comment]
There may be an existing bug for this already, I think we'll need to document our supported contract for whether full-Flutter apps can go into Xcode/gradle and build the app from scratch without ever invoking any Flutter tool commands. 

Beyond that, perhaps Flutter clean might need to do one additional step of deleting `pod_inputs.fingerprint` used in `processPodsIfNeeded` to make sure that a next build does trigger `pod install`.

**Comment by jmagman** [comment]
> There may be an existing bug for this already, I think we'll need to document our supported contract for whether full-Flutter apps can go into Xcode/gradle and build the app from scratch without ever invoking any Flutter tool commands.

Do you mean after a `clean`?  Or on creation?

> Beyond that, perhaps Flutter clean might need to do one additional step of deleting `pod_inputs.fingerprint` used in `processPodsIfNeeded` to make sure that a next build does trigger `pod install`.

That's a good idea!

**Comment by xster** [comment]
> Do you mean after a clean? Or on creation?

I somewhat suspect that there'll be less cognitive load on us the less states there are (and merge those 2 semantically). But I might be missing something.

**Comment by jmagman** [comment]
On mac development the entire "ephemeral" directory gets deleted on `clean`, it won't build in Xcode unless you `flutter run` again.  And iOS add-to-app the entire Xcode project gets deleted, so you definitely need to run `flutter run` again.

**Comment by jmagman** [comment]
> Beyond that, perhaps Flutter clean might need to do one additional step of deleting `pod_inputs.fingerprint` used in `processPodsIfNeeded` to make sure that a next build does trigger `pod install`.

I just realized `flutter clean` already does this since that fingerprint file is in the `build` directory, which gets completely deleted during a clean, and it does run `pod install` after every clean in my tests.
